File: Playlist#2/TokoFD/app/src/main/java/com/example/tokofd/HomeFragment.java
package com.example.tokofd;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.tokofd.model.Product;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HomeFragment extends Fragment {

    private RecyclerView recyclerView;
    private ProductAdapter adapter;
    private List<Product> productList;
    private List<String> productIdList;

    private DatabaseReference produkRef;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_home, container, false);

        recyclerView = rootView.findViewById(R.id.productRecycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        productList = new ArrayList<>();
        productIdList = new ArrayList<>();
        adapter = new ProductAdapter(getContext(), productList, productIdList);
        recyclerView.setAdapter(adapter);

        produkRef = FirebaseDatabase.getInstance().getReference("produk");
        loadProductsFromFirebase();

        return rootView;
    }

    private void loadProductsFromFirebase() {
        produkRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                productList.clear();
                productIdList.clear();

                for (DataSnapshot snap : snapshot.getChildren()) {
                    Product p = snap.getValue(Product.class);
                    if (p != null) {
                        productList.add(p);
                        productIdList.add(snap.getKey());  // Simpan ID produk asli dari database
                    }
                }
                adapter.notifyDataSetChanged();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                Toast.makeText(getContext(), "Gagal memuat produk", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: Playlist#2/Blogging/app/src/main/java/com/example/blogging/HomeFragment.kt
package com.example.blogging

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.blogapp.model.Blog
import com.example.blogging.adapter.BlogAdapter
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class HomeFragment : Fragment() {
    private lateinit var blogRecycler: RecyclerView
    private val blogs = mutableListOf<Blog>()
    private lateinit var adapter: BlogAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_home, container, false)
        blogRecycler = view.findViewById(R.id.blogRecyclerView)
        blogRecycler.layoutManager = LinearLayoutManager(requireContext())
        adapter = BlogAdapter(blogs, requireContext())
        blogRecycler.adapter = adapter
        loadBlogs()
        return view
    }

    private fun loadBlogs() {
        val blogsRef = FirebaseDatabase.getInstance().getReference("blogs")
        blogsRef.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                blogs.clear()
                for (child in snapshot.children) {
                    val blog = child.getValue(Blog::class.java)
                    if (blog != null) {
                        // Set id key dari child key
                        blog.id = child.key ?: ""
                        blogs.add(blog)
                    }
                }
                adapter.notifyDataSetChanged()
            }

            override fun onCancelled(error: DatabaseError) {
                Toast.makeText(requireContext(), "Gagal muat blog: ${error.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }
}
